### ISSUE TITLE
Unexpected nullability warnings for `switch` expressions and `switch` statements utilizing patterns.

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#82101]
```
#nullable enable

interface I1
{
    C2? F { get; }
}

class C2
{
    public string? S = null;
}

interface I2 : I1
{
}

class C11;
class C12;
class C13 : C12, I1
{
    public C2? F => null;
}
class C14 : I2
{
    public C2? F => null;
}
class C15 : I1
{
    public C2? F => null;
}
class C16 : C12, I2
{
    public C2? F => null;
}

class Program
{
    static int Test1((object, int) u)
    {
        return u switch 
        {
            (C12 and I1 and { F: { S: "1" and var x1 } }, 2) => x1.Length ,
            (I2 and { F: { S: ("1" or "2") and var x2 } }, 1) => x2.Length,
            _ => -100
        };
    }   

    static int Test2((object, int) u)
    {
        return u switch 
        {
            (C12 and I1 and { F: { S: "1" } x3 }, 2) => x3.S.Length ,
            (I2 and { F: { S: ("1" or "2") } x4 }, 1) => x4.S.Length,
            _ => -100
        };
    }   

    static int Test3((object, int) u)
    {
        if (u is (C12 and I1 and { F: { S: "1" and var x1 } }, 2)) return x1.Length;
        if (u is (I2 and { F: { S: ("1" or "2") and var x2 } }, 1)) return x2.Length;

        return -100;
    }   

    static int Test4((object, int) u)
    {
        if (u is (C12 and I1 and { F: { S: "1" } x3 }, 2)) return x3.S.Length;
        if (u is (I2 and { F: { S: ("1" or "2") } x4 }, 1)) return x4.S.Length;

        return -100;
    }   

    static int Test5((object, int) u)
    {
        switch (u)
        {
            case (C12 and I1 and { F: { S: "1" and var x1 } }, 2): return x1.Length;
            case (I2 and { F: { S: ("1" or "2") and var x2 } }, 1): return x2.Length;
        }

        return -100;
    }   

    static int Test6((object, int) u)
    {
        switch (u)
        {
            case (C12 and I1 and { F: { S: "1" } x3 }, 2): return x3.S.Length;
            case (I2 and { F: { S: ("1" or "2") } x4 }, 1): return x4.S.Length;
        }

        return -100;
    }   
}
```

Observed:
```
// (43,66): warning CS8602: Dereference of a possibly null reference.
//             (I2 and { F: { S: ("1" or "2") and var x2 } }, 1) => x2.Length,
Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x2").WithLocation(43, 66),
// (53,58): warning CS8602: Dereference of a possibly null reference.
//             (I2 and { F: { S: ("1" or "2") } x4 }, 1) => x4.S.Length,
Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x4").WithLocation(53, 58),
// (53,58): warning CS8602: Dereference of a possibly null reference.
//             (I2 and { F: { S: ("1" or "2") } x4 }, 1) => x4.S.Length,
Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x4.S").WithLocation(53, 58),
// (79,76): warning CS8602: Dereference of a possibly null reference.
//             case (I2 and { F: { S: ("1" or "2") and var x2 } }, 1): return x2.Length;
Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x2").WithLocation(79, 76),
// (90,68): warning CS8602: Dereference of a possibly null reference.
//             case (I2 and { F: { S: ("1" or "2") } x4 }, 1): return x4.S.Length;
Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x4").WithLocation(90, 68),
// (90,68): warning CS8602: Dereference of a possibly null reference.
//             case (I2 and { F: { S: ("1" or "2") } x4 }, 1): return x4.S.Length;
Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x4.S").WithLocation(90, 68)
```

Expected:
No warnings as none reported in `Test3` and `Test4`.

**Comment by madelyngamble2** [comment]
@AlekseyTs May I resolve this issue?

**Comment by AlekseyTs** [comment]
@madelyngamble2 

> May I resolve this issue?

If you would like to submit a PR with a fix, that would be fine with me. However, consider coordinating with @RikkiGibson, to whom this issue is currently assigned.

**Comment by RikkiGibson** [comment]
@madelyngamble2 Feel free to work on this, and let us know any questions you have as you investigate the issue, either here or in the [C# Discord](http://aka.ms/discord-csharp-roslyn).

**Comment by madelyngamble2** [comment]
@RikkiGibson My approach will be

1. Extract Pattern Nullability Analysis
  Create a reusable method that analyzes patterns and determines which variables/properties are guaranteed non-null
  Handle property patterns, or patterns, variable bindings, etc.
2. Update Switch Expression Analysis
  Modify VisitSwitchExpression in NullableWalker.cs
  Apply pattern nullability information to each switch arm
  Mark bound variables as non-null when patterns guarantee it
3. Update Switch Statement Analysis
 Similar changes to VisitSwitchStatement
 Ensure consistency with switch expressions

Regards
